### PR TITLE
Synaptics MST panamera support

### PR DESCRIPTION
**Pull request by superm1**
In an effort to move https://github.com/hughsie/fwupd/pull/560 along from a bazillion nits and some minor logic errors I've cleaned it up myself in this PR.

I kept the original commits from @ryanchang0624 (squashed into one) and then iterated on top of them a bit.

I haven't tested this yet on panamera myself but I should be able to more easily after #602 is merged.

@ryanchang0624 if there is anything I broke in the process of cleaning your PR up please speak up.  If you can please test it on EVB tesla/leaf and panamera I would really appreciate it.

### CODE REVIEW

**Comment by hughsie** [comment]
I've fixed up a bit more, coverity also seems happy now. If you do one more pass for anything obvious then this LGTM once the other PR is fixed up.

**Comment by superm1** [comment]
I'm still doing some testing with it (found at least that one logic error) and need to double check to make sure last generation didn't break.

Once it's ready how do you want to merge this?  I am tending to think rather than our regular "squash and merge" or "rebase and merge" we should actually include a merge commit with this since @ryanchang0624 contributed a ton here and I don't want to lose attribution in a squash.

**Comment by hughsie** [comment]
I think squash is fine, we can just include all the authors in the commit msg. Sometimes I just set the author to the person that did the initial work and then add my signoff -- on the logic it's not important to me but might be important to another person.

**Comment by superm1** [comment]
OK, I've tested that this branch works properly on Panamera now and squashed it down.  I also had one successful flash on last generation, but also one failure.  I think they were cascaded though, so I need to investigate a little more closely.

Otherwise the branch is in good shape in my mind.

@ryanchang0624 a few questions for you if you can answer:

1. I notice there is no MST reboot command at the end of the flash process over DP aux channel.  Is that intentional?
2. I notice that the Panamera flash bank tag header is being written with _current_ date information rather than the _build_ information which was already in the file.  Is this what you expect to happen?  If it's supposed to be _build_ information then we probably should just strip that tag writing section entirely.

I don't think we need to block merging this on that information, these can be follow ups if need be.

**Comment by superm1** [comment]
I think it was a bad payload (I manually extracted it from the EXE).  Flashing something older that I have the bin file worked out properly.

**Comment by ryanchang0624** [comment]
Hi Mario,

Please see my comments below.
Regards,
Ryan

**Comment by superm1** [comment]
@ryanchang0624  how about that second question about tags?

**Comment by ryanchang0624** [comment]
Hi Mario,


  1.  I notice that the Panamera flash bank tag header is being written with current date information rather than the build information which was already in the file. Is this what you expect to happen? If it's supposed to be build information then we probably should just strip that tag writing section entirely.

I check the source code of VmmDPTool,  it is current time for sure.
Do you have any concern about this? I could check with Leo.

Regards,
Ryan

**Comment by superm1** [comment]
> > [Ryan] You are right. The updating flow is came from VmmDpTool and there is a button “reset board” in GUI. So, usually user will manually perform reboot. But for fwupd, I agreed that it should automatically reboot after updating.

OK can you please follow up with adding reboot support in a future PR after this one is merged?

> I check the source code of VmmDPTool,  it is current time for sure.
>Do you have any concern about this? I could check with Leo.

It just didn't match the discussions I had about tags and it makes it more difficult to verify if a bank is already up to date or not since you can't checksum it immediately.  If that's the flow that should be used that's fine though.  I guess please check with others at Synaptics to confirm and we'll plan to keep it as is.

**Comment by superm1** [comment]
@hughsie That last one might make sense to generally do for benchmarking all plugins performance but I wanted to run it by you.  I've got a particular need on synaptics that might not exist for other plugins.

**Comment by superm1** [comment]
@ryanchang0624 I think that 309ed91 does reboot correctly (seems functional at least), but if you can please double check it looks right from your side.  I've not done it over DP aux interface before.

**Comment by superm1** [comment]
I've done a lot of testing today across both panamera and an older tesla/leaf board.  Panamera seems solid  including reboot and jumps up and down from ESM.

Surprisingly I ran into problems with the older tesla/leaf board, and I think they were actually issues that existed previously and not introduced by adding this new support.  Problems ranged from:

- Invalid board ID after flash
- Checksum invalid "during" flash
- Failed "previous" runs couldn't recover (remote control register still enabled).
- Racing with Dell plugin to initialize dock type string (and frequently losing).

I've added various commits for these things and now haven't been able to break tesla/leaf in any creative ways with many upgrade and downgrade cycles.

**Comment by superm1** [comment]
Also, re-ordered them so the commit adding panamera support is more distinct from the rest of the changes that have happened.  I think many of those are applicable to bring to `1_0_X` too since I was able to get some boards in very bad states without them (eg `0xFFFF` `priv->board_id`), but @hughsie can make the call on that.

File: plugins/synapticsmst/fu-plugin-synapticsmst.c [hunk 1]
```diff
@@ -354,6 +355,9 @@ fu_plugin_update (FuPlugin *plugin,
 	}
 	fu_device_set_version (dev, synapticsmst_device_get_version (device));
 
+	g_debug ("Updating %s took %f seconds", fu_device_get_name (dev),
```
Comment (hughsie) [inline plugins/synapticsmst/fu-plugin-synapticsmst.c]: this seems like the thing we ought to do in the daemon

**Comment by hughsie** [comment]
LGTM, although I do think the `GTimer` bit would be useful for all plugins.

**Comment by superm1** [comment]
@ryanchang0624 it's been merged to master so if you can please verify with evb on your side that would be appreciated. I did tests from Dell products with Panamera and Leaf, things looked good to me.

Regarding tag - my main concern is that we stay consistent with Windows and Linux. So please let your colleagues discuss with appropriate decision makers and go from there. This implementation is fine if it's going to continue to match windows tool approach.

**Comment by superm1** [comment]
@hughsie those first 4 commits, do they seem acceptable to you for 1_0_X?

**Comment by hughsie** [comment]
> do they seem acceptable to you for 1_0_X

Yup, feel free to backport those; basically anything for hardware enablement that isn't going to break things :)

**Pull request merged**